bubble_sortアルゴリズム作成

kawanaka-s-ts opened the pull request:
## 課題

バブルソートを行うアルゴリズムをRubyで記述する。

## このPRがマージされると？

バブルソートが行えるようになる。

## 開発方針（このPRで具体的に修正する箇所）

 - 繰り返しには、whileを使用する。 

## TODO

- [x] 要素の入れ替え方法を、復習した。
- [x] 繰り返し処理の中に、繰り返し処理がある場合のコードの書き方、考え方を学習した。

## テスト結果とテスト項目

- [x] 隣り合った要素で大小を比較する　要素を昇順に並び替えた配列を返す。

## 今回保留した項目



## 補足

## PRマージまでのチェックリスト

- [ ] merge準備完了
  - [ ] この項目以外のチェックボックス全てにチェックが入っている
  - [ ] `コードレビュー済` ラベルが付与されている
  - [ ] conflictが発生していない
  - [x] 関連するテストコードが存在する
  - [x] 関連するテストコードが全て通っている

kawanaka-s-ts committed:
bubble_sortアルゴリズム作成

kawanaka-s-ts committed:
fixup! bubble_sortアルゴリズム作成

- コメント誤字修正

kawanaka-s-ts committed:
fixup! bubble_sortアルゴリズム作成

kawanaka-s-ts committed:
テスト時はランダムに並び替えたnumbersを渡すように変更

kyakumoto-ts commented on spec/bubble_sort_spec.rb:
```diff
@@ -0,0 +1,15 @@
+
+describe BubbleSort do
+  describe "#sort" do
+    context "隣り合った要素で大小を比較する" do
```
##### L8

`隣り合った要素で大小を比較しない場合` が存在しないので、 `context` を設定する必要はないと思います。

kawanaka-s-ts replied:
修正しました。

kyakumoto-ts commented on lib/bubble_sort.rb:
```diff
@@ -0,0 +1,31 @@
+        # 隣り合った要素で大小を比較して、右の要素の方が小さい時
+        if numbers[index] > numbers[index + 1]
+          # 要素を入れ替える
+          numbers[index], numbers[index + 1] = numbers[index + 1], numbers[index]
```
##### L16

こういう場合は2行にしてしまった方が良いと思います。
逆に配列を展開する様な場面では1行で書いた方がスッキリする気がします（個人的には、あまりこういう場面に出くわしたことはないですが）。
```
items = [1, 2, 3]
a, b, c = *items
a #=> 1
b #=> 2
c #=> 3

def return_array
  [4, 5, 6]
end
a, b, c = *return_array
a #=> 4
b #=> 5
c #=> 6
```

kawanaka-s-ts replied:
場面ごとに使い分けを考えれるようにしていきます。
修正しました。

kyakumoto-ts commented on lib/bubble_sort.rb:
```diff
@@ -0,0 +1,31 @@
+class BubbleSort
+  # sortメソッドここから
+  def sort(numbers)
```
##### L3

`BubbleSort` というクラス名が並び替えを表しているので、 `#sort` というメソッド名はやや冗長であり、また曖昧かなと思います。
昇順で並び替える様になっていますので、その様なメソッド名の方が望ましいと思います。

kawanaka-s-ts replied:
昇順なので、`ascending_order`という名前にしてみました。

kyakumoto-ts reviewed (changes requested):
内部実装とテストについてです。

kawanaka-s-ts committed:
fixup! bubble_sortアルゴリズム作成

- 不要なcontext削除

kawanaka-s-ts committed:
fixup! bubble_sortアルゴリズム作成

- メソッド名変更

kawanaka-s-ts committed:
fixup! bubble_sortアルゴリズム作成

- 要素の入れ替え部分を変更（2行に分けた）
- 配列を展開するような場合は、1行の方が良い

kyakumoto-ts commented on lib/bubble_sort.rb:
```diff
@@ -1,6 +1,6 @@
-  # sortメソッドここから
-  def sort(numbers)
+  # ascending_orderメソッドここから
+  def ascending_order(numbers)
```
##### L3

良いと思います！
`asc` としてしまっても、意味が通じると思います。

kawanaka-s-ts replied:
`asc`に変更しました。

kyakumoto-ts reviewed (commented):
```
fixup! bubble_sortアルゴリズム作成
```

kyakumoto-ts reviewed (changes requested):
```
fixup! bubble_sortアルゴリズム作成
```
すみません、間違えてコメントなしになってしまいました。。。

kawanaka-s-ts committed:
fixup! fixup! bubble_sortアルゴリズム作成

- コメント修正

kyakumoto-ts commented on lib/bubble_sort.rb:
```diff
@@ -13,7 +13,8 @@ def ascending_order(numbers)
           # 要素を入れ替える
-          numbers[index], numbers[index + 1] = numbers[index + 1], numbers[index]
+          numbers[index], numbers[index + 1] =
+          numbers[index + 1], numbers[index]
```
##### L16, L17

2行に分けるのであれば、1つずつ代入する方が可読性が高いです。
また、 `=` の右辺を改行する場合はインデントを下げるべきです。

kawanaka-s-ts replied:
変数に値を保持して、入れ替えを行うように変更しました。
また、比較の際に使う`index + 1`は変数jに代入するように変更しました。

shugo replied:
変数が必要なのと3行になってしまっているのが元のコードと比べてどちらがよいか悩ましいところですね。

Rubyでswapを行うのに変数の導入を避けて多重代入を使うのはわりと一般的だと思いますが、多重代入を避けるなら

```
numbers[index..j] = numbers[index..j].reverse
```

といった書き方を考えました。ただ、読みやすいかどうかは微妙かもしれません。
右辺を `numbers[j..index]` とは書けないのですよね。

kyakumoto-ts replied:
>Rubyでswapを行うのに変数の導入を避けて多重代入を使うのはわりと一般的だと思いますが、多重代入を避けるなら

多重代入でswapができるのを自分が知っていなかったです。
`swap` というものを覚えて欲しかったというのもあったので、目的は達成できました笑

>変数が必要なのと3行になってしまっているのが元のコードと比べてどちらがよいか悩ましいところですね。
>
>numbers[index..j] = numbers[index..j].reverse
>
>といった書き方を考えました。ただ、読みやすいかどうかは微妙かもしれません。
>
>右辺を `numbers[j..index]` とは書けないのですよね。

```
numbers[index..j] = numbers[j..index]
```
確かに、こう書けたらぱっと見で反対になってる感は伝わりますね。

変数名次第で 多重代入、改行、どちらもアリでありナシである という感じがしますね。
ですが、余分な変数使わないで済む多重代入の方に軍配があがるかなーという感想です。

kyakumoto-ts replied:
swapってメソッドは用意されていないのですね。ちょっと意外でした。

shugo replied:
そうですねえ、一長一短ですね。

多重代入も内部的には配列で実現されているので無駄はあるんですが、コードの字面上は余分な変数がいらなくなる利点はありますね。

とくに `x, y = y, x`のように単純な例だとすっきりします。

kyakumoto-ts reviewed (changes requested):
```
fixup! bubble_sortアルゴリズム作成
```

kawanaka-s-ts committed:
fixup! fixup! fixup! bubble_sortアルゴリズム作成

- メソッド名変更ascending→asc

kawanaka-s-ts committed:
要素の入れ替え方を変更

-
そのまま入れ替えようとすると、元の値が書き換わってしまう為、入れ替えがうまくいかない。
-
そのため値を保持する変数を用意して、そこに値を退避させておき後から戻す。

-　当初のやり方は、値を同時に入れ替えるので変数を用意する必要はない。

kawanaka-s-ts committed:
fixup! 要素の入れ替え方を変更

- 変数jにindex + 1を代入するように変更

shugo commented on spec/bubble_sort_spec.rb:
```diff
@@ -0,0 +1,13 @@
+require "spec_helper"
+require "bubble_sort"
+
+numbers = [*1..10].sort_by{rand}
```
とくに乱数アルゴリズムを指定したいといったことがなければ、 `[*1..10].shuffle` とするとより簡潔に記述できます。

kawanaka-s-ts replied:
確かにshuffleの方が簡潔ですね。
修正します。

kawanaka-s-ts committed:
UFO演算子を用いて昇順、降順を切り替えれるように変更

- UFO演算子の役割について理解が深まった
-
処理の順序は、（）で囲む事で数式のような感じに実行する順序を変更する事ができる

kawanaka-s-ts committed:
アルゴリズムが違うパターン作成

kawanaka-s-ts committed:
asc_orderメソッド、desc_orderメソッド追加

shugo reviewed (approved):
修正を確認しました :+1:

kyakumoto-ts commented on lib/bubble_sort.rb:
```diff
@@ -33,6 +33,24 @@ def change_order(numbers, pattern)
+    # patternに1を代入（昇順）
+    pattern = 1
+    # sortメソッド呼び出し
+    sort(numbers, pattern)
```
##### L41 - L44

L41の `pattern` の様な変数を、説明変数と呼んだりします（正式な名称かは分からないですが、そうやって呼んでいる人が一定数いそうです）。
代入せずに値を直接使っても良い場面で、変数に代入することで何のための変数かを分かりやすくする感じです。

個人的には使うべき時と使わなくて良い時があると考えていて、今回で言えば後者です。
ですので、 L44は下記の様に書いて良いと思います。
```
sort(numbers, 1)
```

こうした時の `1（数値や文字列etc）` は、マジックナンバーと呼ばれたりします。
`値が直接書いてあり、どういう意味があるかが分からない値` のことです。
マジックナンバーは避けた方が良いのですが、今回の場合は

1. メソッド名が `昇順` であることを説明してくれている
1.  `#sort` 内で条件判定をしているのが `<=>` である（-1, 0, 1 のどれかが返る）
1. 上記より、 `-1, 0, 1` という値を利用することが前提になる
1. メソッド数が少なく、#sortの利用が限定的である

という理由から、必要十分だろうと考えているためです。

ですが、今回の様なパターンでそれでもちょっと分かりづらいとなれば、説明変数でなく定数を使用すると思います。
クラス内で宣言をしておく感じです。
```
ORDER_BY_ASC = 1
ORDER_BY_DESC = -1

def asc_order(numbers)
  ...
  sort(numbers, ORDER_BY_ASC)
end
```

書いていて、どっちが良いのか分からなくなってきた。。。

kyakumoto-ts replied:
定数を利用した場合のメリットとしては、 `<=>` の処理を置き換える際に、値がなんだったのかを気にしなくても良い（ORDER_BY_ASCの中身だけを書き換えれば良い）というのもあります。

kyakumoto-ts replied:
2箇所で `ORDER_BY_ASC` を利用する必要があるなら迷わず定数利用とかになると思いますが、今回はその必要ってあるのかな？という感じです。

shugo replied:
私もこのケースでは `sort(numbes, 1)` でよいように思いますが、どちらかというとsortの第2引数の意味がわかりにくいことの方に問題があるように思います（だから `pattern = 1` と書きたくなったのではないでしょうか）。

例えば、 `sort(numbers, pattern: 1)` のようにキーワード引数を使うと引数の意味をわかりやすくできますし、もっと言えば1というのはsortの実装の都合なので、 `sort(numbers, order: :asc)` のようなインタフェースにした方がユーザにとってはわかりやすいと思います。
また、組み込みクラスだと<=>は1, 0, -1を返しますが、ユーザ定義クラスだと正、0、負を返すかもしれません（そうすると `Foo#<=>(other)` を `self - other` として実装できて楽なので）、0, 1, -1以外の値が<=>で返されても動くようにsortの実装についても見直した方がよいかもしれません。

kyakumoto-ts replied:
ありがとうございます！

>私もこのケースでは `sort(numbes, 1)` でよいように思いますが、どちらかというとsortの第2引数の意味がわかりにくいことの方に問題があるように思います（だから `pattern = 1` と書きたくなったのではないでしょうか）。
>
>例えば、 `sort(numbers, pattern: 1)` のようにキーワード引数を使うと引数の意味をわかりやすくできますし、もっと言えば1というのはsortの実装の都合なので、 `sort(numbers, order: :asc)` のようなインタフェースにした方がユーザにとってはわかりやすいと思います。
また、組み込みクラスだと<=>は1, 0, -1を返しますが、ユーザ定義クラスだと正、0、負を返すかもしれません（そうすると `Foo#<=>(other)` を `self - other` として実装できて楽なので）、0, 1, -1以外の値が<=>で返されても動くようにsortの実装についても見直した方がよいかもしれません。

キーワード付き引数とシンボルを利用すれば、確かに定数も要らないですね。
シンボルを利用した場合ですが、 `<=>` をユーザー定義クラスに置き換える必要があるという認識なのですが、正しいでしょうか？

shugo replied:
> シンボルを利用した場合ですが、 <=> をユーザー定義クラスに置き換える必要があるという認識なのですが、正しいでしょうか？

いや、シンボルを直接 `<=>` に渡すことは考えていませんでした。
色々やり方は考えられますが、例えば以下のようなイメージです。

```
def sort(numbers, order: :asc)
  cmp = order == :asc ? :> : :<
  ...
    if numbers[index].send(cmp, numbers[j])
```

kyakumoto-ts replied:
なるほどです！ありがとうございます！

kawanaka-s-ts committed:
fixup! asc_orderメソッド、desc_orderメソッド追加

kawanaka-s-ts committed:
fixup! fixup! asc_orderメソッド、desc_orderメソッド追加

- letを使ってテストコードの重複を修正
